### PR TITLE
Inline Arg Optimization

### PR DESCRIPTION
**Pull request by yullin-ms**
This is to do inline arg optimization when we do argument length and argument constant optimizations

### CODE REVIEW

File: lib/Backend/GlobOptBailOut.cpp [hunk 1]
```diff
@@ -482,6 +482,16 @@ GlobOpt::CaptureByteCodeSymUses(IR::Instr * instr)
 void
 GlobOpt::ProcessInlineeEnd(IR::Instr* instr)
 {
+    if (instr->m_func->hasArgLenAndConstOpt && instr->m_func->unoptableInlineArgCount == 0)
+    {
+        instr->m_func->hasUnoptimizedArgumentsAccess = false;
+        if (DoInlineArgsOpt(instr->m_func))
+        {
+            instr->m_func->m_hasInlineArgsOpt = true;
+            instr->m_func->frameInfo = instr->m_func->cachedInlineeFrameInfo;
```
Comment (sigatrev) [inline lib/Backend/GlobOptBailOut.cpp]: is it possible to have frameInfo non-null here? is it possible to have cacheInlineeFrameInfo null here? If either of those situations would be bad we should add an assert.

File: lib/Backend/GlobOptBailOut.cpp [hunk 2]
```diff
@@ -570,23 +580,34 @@ GlobOpt::TrackCalls(IR::Instr * instr)
...
+    {
         Assert(instr->m_func->GetParentFunc() == this->currentBlock->globOptData.curFunc);
         Assert(instr->m_func->GetParentFunc());
         this->currentBlock->globOptData.curFunc = instr->m_func;
 
         this->func->UpdateMaxInlineeArgOutSize(this->currentBlock->globOptData.inlinedArgOutSize);
         this->EndTrackCall(instr);
 
-        if (DoInlineArgsOpt(instr->m_func))
+        auto createFrameInfo = [&](Func* inlineeFunc)
```
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/GlobOptBailOut.cpp]: Thanks, I didn't know that. I'll update it to reflect that.

File: lib/Backend/GlobOpt.cpp [hunk 3]
```diff
@@ -13208,10 +13211,10 @@ GlobOpt::OptArgLenAndConst(IR::Instr* instr, Value** src1Val)
     if (instr->usesStackArgumentsObject && instr->IsInlined())
     {
         IR::Opnd* src1 = instr->GetSrc1();
-        auto replaceInstr = [&](IR::Opnd* newopnd)
+        auto replaceInstr = [&](IR::Opnd* newopnd, Js::OpCode opCode = Js::OpCode::Ld_A)
```
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: I can modify it to just pass in the opcode.

File: lib/Backend/GlobOptBailOut.cpp (hunk 1 above)
Comment (yullin-ms) [reply lib/Backend/GlobOptBailOut.cpp]: This may be possible, would a check before entering this scope be better than an assert?

Comment (yullin-ms) [reply lib/Backend/GlobOptBailOut.cpp]: 12d2aa8e07c27501ef78db203fe04a8136392566

File: lib/Backend/GlobOpt.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: 12d2aa8e07c27501ef78db203fe04a8136392566

File: lib/Backend/GlobOptBailOut.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yullin-ms) [reply lib/Backend/GlobOptBailOut.cpp]: 12d2aa8e07c27501ef78db203fe04a8136392566

File: lib/Backend/Func.cpp [hunk 4]
```diff
@@ -66,6 +66,7 @@ Func::Func(JitArenaAllocator *alloc, JITTimeWorkItem * workItem,
     m_hasCalls(false),
     m_hasInlineArgsOpt(false),
     m_canDoInlineArgsOpt(true),
+    unoptableInlineArgCount(0),
```
Comment (rajatd) [inline lib/Backend/Func.cpp]: could you rename this to, say, unoptimizableArgumentsObjReference ?

File: lib/Backend/GlobOpt.cpp [hunk 5]
```diff
@@ -13252,12 +13261,23 @@ GlobOpt::OptArgLenAndConst(IR::Instr* instr, Value** src1Val)
                     {
                         IR::Opnd * undefined = IR::AddrOpnd::New(instr->m_func->GetScriptContextInfo()->GetUndefinedAddr(), IR::AddrOpndKindDynamicVar, instr->m_func, true);
                         undefined->SetValueType(ValueType::Undefined);
-                        replaceInstr(undefined);
+                        replaceInstr(undefined, Js::OpCode::Ld_A);
```
Comment (rajatd) [inline lib/Backend/GlobOpt.cpp]: Isn't this incorrect for TypeofElem?

Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: Should it still be TypeOf opcode for undefined, does that not return undefined?

Comment (rajatd) [reply lib/Backend/GlobOpt.cpp]: the string "undefined" not the object undefined

---
In reply to: [222496035](https://github.com/Microsoft/ChakraCore/pull/5733#discussion_r222496035) [](ancestors = 222496035)

Comment (yullin-ms) [reply lib/Backend/GlobOpt.cpp]: Gotcha, I have made the modifications for that.

**Pull request merged**